Add a support for standard SSL configurations, proxy and basic auth configs.

mashhurs opened the pull request:
## Description
  - Add SSL/TLS support for HTTPS schema registry connections
    - Add `ssl_enabled` option to enable/disable SSL
    - Add `ssl_certificate` and `ssl_key` options for PEM-based client authentication (unencrypted keys only)
    - Add `ssl_certificate_authorities` option for PEM-based server certificate validation
    - Add `ssl_verification_mode` option to control SSL verification (full, none)
    - Add `ssl_cipher_suites` option to configure cipher suites
    - Add `ssl_supported_protocols` option to configure TLS protocol versions (TLSv1.1, TLSv1.2, TLSv1.3)
    - Add `ssl_truststore_path` and `ssl_truststore_password` options for server certificate validation (JKS/PKCS12)
    - Add `ssl_keystore_path` and `ssl_keystore_password` options for mutual TLS authentication (JKS/PKCS12)
    - Add `ssl_truststore_type` and `ssl_keystore_type` options (JKS or PKCS12)
  - Add HTTP proxy support with `proxy` option
  - Add HTTP basic authentication support with `username` and `password` options

Integration test environment is same as https://github.com/logstash-plugins/logstash-integration-kafka

mashhurs committed:
Add a support for standard SSL configurations, proxy and basic auth configs.

mashhurs committed:
Increase test coverage with unit tests.

mashhurs committed:
Separate remote schema registry tests from encode and decode.

mashhurs committed:
Fix CI failures.

mashhurs committed:
Prepare kafka and confluent schema registry environment to use in the integration tests.

mashhurs committed:
Basic integration test with schema registry.

mashhurs committed:
Add integration tests for SSL and auth based schema registry. Add integration steps in the Travis.

mashhurs committed:
Exclude 7.current from travis integration jobs.

mashhurs commented on .ci/run.sh:
```diff
@@ -0,0 +1,34 @@
+if [[ "$INTEGRATION" != "true" ]]; then
+  bundle exec rake test
+else
+  # Define the Kafka:Confluent version pairs
```
Using same as https://github.com/logstash-plugins/logstash-integration-kafka/blob/main/.ci/run.sh to keep compatibility

mashhurs commented on .travis.yml:
```diff
@@ -1,2 +1,11 @@
+    - env: INTEGRATION=true LOG_LEVEL=info ELASTIC_STACK_VERSION=9.current
+    - env: INTEGRATION=true SNAPSHOT=true LOG_LEVEL=info ELASTIC_STACK_VERSION=8.current
+    - env: INTEGRATION=true SNAPSHOT=true LOG_LEVEL=info ELASTIC_STACK_VERSION=9.current
+    - env: INTEGRATION=true SNAPSHOT=true LOG_LEVEL=info ELASTIC_STACK_VERSION=main
```
Excluded 7.current as I do see we have failures with kafka & JDK incompatibility. There might be a possibility but for the given EOL 7.x, I am skipping it for now unless there is a strong feedback. 
```
```
      Picked up _JAVA_OPTIONS: -Djava.net.preferIPv4Stack=true

      Error: LinkageError occurred while loading main class kafka.tools.StorageTool

        java.lang.UnsupportedClassVersionError: kafka/tools/StorageTool has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0

      + build/kafka/bin/kafka-storage.sh format --cluster-id --config build/kafka/config/server.properties --ignore-formatted --standalone

      Picked up _JAVA_OPTIONS: -Djava.net.preferIPv4Stack=true

      Error: LinkageError occurred while loading main class kafka.tools.StorageTool

        java.lang.UnsupportedClassVersionError: kafka/tools/StorageTool has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0

      ```
```

mashhurs commented on docs/index.asciidoc:
```diff
@@ -83,9 +83,25 @@ output {
+| <<plugins-{type}s-{plugin}-ssl_verification_mode>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-tag_on_failure>> |<<boolean,boolean>>|No
 | <<plugins-{type}s-{plugin}-target>> |<<string,string>>|No
+| <<plugins-{type}s-{plugin}-username>> |<<string,string>>|No
```
Identical with es-output plugin SSL options: https://www.elastic.co/docs/reference/logstash/plugins/plugins-outputs-elasticsearch#plugins-outputs-elasticsearch-options

mashhurs commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+    {:user => @username, :password => @password.value}
+  end
+
+  def validate_ssl_settings!
```
SSL validation logics are same with `elastic_integration` (in the past we reviewed with Rye) - https://github.com/elastic/logstash-filter-elastic_integration/blob/main/lib/logstash/filters/elastic_integration.rb#L234

mashhurs commented on spec/unit/avro_spec.rb:
```diff
@@ -0,0 +1,881 @@
+    end
+  end
+
+  context "remote schema registry" do
```
Re-organized spec folder structure and moved this file into unit folder.
Above changes are not touched, please review after this point.

mashhurs commented on spec/unit/resources/do_not_remove_path/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Empty dir for test cases to run, imitates readable/non-readable/writable folder
+# When configs are :path validated, existed path is required
```
As comment idicates!

kaisecheng commented on spec/integration/kafka_test_setup.sh:
```diff
@@ -0,0 +1,80 @@
+echo "authentication.method=BASIC" >> "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
+echo "authentication.roles=admin,developer,user,sr-user" >> "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
+echo "authentication.realm=SchemaRegistry-Props" >> "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
+cp fixtures/jaas.config "build/confluent_platform/etc/schema-registry"
```
if the plugin needs to test against schema registry confluent v7 and v8 to ensure compatibility, this script need to add [zookeeper](https://github.com/logstash-plugins/logstash-integration-kafka/blob/main/kafka_test_setup.sh#L28-L30) and [jaas3.config](https://github.com/logstash-plugins/logstash-integration-kafka/blob/main/kafka_test_setup.sh#L81-L83)

mashhurs replied:
In case of SSL connection and basic auth I am implementing, is there differences between v7 and v8?
As I went through [7.9](https://docs.confluent.io/platform/7.9/schema-registry/installation/config.html) and [8.1](https://docs.confluent.io/platform/8.1/schema-registry/installation/config.html) versions, I hadn't see the difference.

kaisecheng replied:
As the script defines the [pairs](https://github.com/logstash-plugins/logstash-codec-avro/pull/47/files#diff-7b436cb5923ba7f66ccf2e140b342330400dcb1d6776d03444564dc34c5379f1R13) including v7 and the [fallback](https://github.com/logstash-plugins/logstash-codec-avro/pull/47/files#diff-1981bbd50b7de373914d07bf06e1694d7c6756b031e886a1e8a029eff4619d6cR249), I was wondering if you intend to test against old version.

I am curious in what situation the fallback is used. In general, it is good to have the CI tests run against the confluent versions that we support. I will let you make the call.

mashhurs replied:
To clarify the fallback, [current `schema_uri`](https://www.elastic.co/guide/en/logstash/8.14/plugins-codecs-avro.html#plugins-codecs-avro-schema_uri) can accept local or remote HTTP URI. If the schema is the let's say in blob/storage (like S3), it can retrieve and use as it is. Basically, it is a plugin backward compatibility.

mashhurs replied:
One example scenario, if you have protobuf type of schema directly used in the S3 cloud storage, which has just body (no metada, etc..), for this case we will not have `schema` field and we use as it is to keep current behavior.

kaisecheng commented on spec/integration/kafka_test_setup.sh:
```diff
@@ -0,0 +1,80 @@
+
+cp "build/confluent_platform/etc/schema-registry/schema-registry.properties" "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
+echo "authentication.method=BASIC" >> "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
+echo "authentication.roles=admin,developer,user,sr-user" >> "build/confluent_platform/etc/schema-registry/authed-schema-registry.properties"
```
fixture/pwd file has no "sr-user" role. You can remove it

mashhurs replied:
Yup, good catch! Removed.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+      run_integration_script("stop_schema_registry.sh")
+      sleep 2
+      run_integration_script("start_schema_registry.sh")
+      sleep 5
```
does it intend to sleep here?

mashhurs replied:
Removed all sleeps and used `Stud.retry`.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+
+    after(:all) do
+      run_integration_script("stop_schema_registry.sh")
+      sleep 2
```
once the script is executed, the registry should finish shutdown and no need to sleep

mashhurs replied:
Removed.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+    end
+  end
+
+  def register_schema(schema_registry_url, schema_json, username: nil, password: nil, ssl_options: {})
```
I don't see any usages of this method

mashhurs replied:
Yup, removed.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+        raise "Schema Registry at #{url} did not become available within #{timeout} seconds after #{attempt} attempts"
+      end
+
+      sleep 2
```
How about `sleep 1`?

kaisecheng replied:
I think Stud.try is a good alternative to replace the whole loop

```ruby
  Stud.try(20.times, [Manticore::SocketException, StandardError, RSpec::Expectations::ExpectationNotMetError]) do
    expect(client.get(url).code).to eq(200)
  end
```

mashhurs replied:
Removed all sleeps and used Stud.retry. Thanks for reminding Stud.

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+    if http_connection
+      ssl_config_provided = original_params.keys.select {|k| k.start_with?("ssl_") && k != "ssl_enabled" }
+      if ssl_config_provided.any?
+        raise_config_error! "When SSL is disabled, the following provided parameters are not allowed: #{ssl_config_provided}"
```
schema_uri => "http://....."
ssl_enabled => false
ssl_truststore_path => "/path/to/truststore"

IMO, the above config is valid. The uri points to http and ssl is explicitly set to false. I would prefer not to raise error

mashhurs replied:
When `ssl_enabled => false`, it makes sense to me accepting _only_ valid configs. For this case, it may sounds the plugin is actually using trustore.

kaisecheng replied:
👌

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+      end
+
+      credentials_configured = @username && @password
+      @logger.warn("Credentials are being sent over unencrypted HTTP. This may bring security risk.") if credentials_configured && http_connection
```
`&& http_connection` is redundant

mashhurs replied:
Yup, removed!

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+    end
+
+    client = Manticore::Client.new(client_options)
+    response = client.get(uri_string).call
```
we need retry for the failure call

mashhurs replied:
Yup, implemented with Stud and 3 times retry. Didn't think about making a separate config for the retry. We can extend later.

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+
+  def fetch_remote_schema(uri_string)
+    client_options = {}
+
+    if @proxy && !@proxy.empty?
+      client_options[:proxy] = @proxy.to_s
+    end
+
+    basic_auth_options = build_basic_auth
+    client_options[:auth] = basic_auth_options unless basic_auth_options.empty?
+
+    if @ssl_enabled
+      ssl_options = build_ssl_options
+      client_options[:ssl] = ssl_options unless ssl_options.empty?
+    end
+
+    client = Manticore::Client.new(client_options)
+    response = client.get(uri_string).call
+
+    unless response.code == 200
+      raise "HTTP request failed: #{response.code} #{response.message}"
+    end
+
+    body = response.body
+    # Response may contain schema metadata, schema field is what we need
+    # Example response: {"subject":"test-no-auth-1763597024","version":1,"id":1,"guid":"5c6c5f26-e876-e5ab-02b0-8d9bebbc90d7","schemaType":"AVRO","schema":"{"type":"record","name":"TestRecord","namespace":"com.example","fields":[{"name":"message","type":"string"},{"name":"timestamp","type":"long"}]}","ts":1763597024561,"deleted":false}
+    begin
+      parsed = JSON.parse(body)
+      if parsed.is_a?(Hash) && parsed.has_key?('schema')
+        parsed['schema']
+      else
+        # fallback to use the response as it is
+        body
+      end
+    rescue JSON::ParserError
+      # Not JSON, return as-is (probably a direct schema)
+      body
+    end
```
we can unwrap begin end
```suggestion
    parsed = JSON.parse(body)
    if parsed.is_a?(Hash) && parsed.has_key?('schema')
      parsed['schema']
    else
      # fallback to use the response as it is
      body
    end
  rescue JSON::ParserError
    # Not JSON, return as-is (probably a direct schema)
    body
```

mashhurs replied:
Applied after confirmation (local tests) with `client = nil`.

kaisecheng reviewed (commented):
I haven't finished the test yet. The main feedbacks are to add retry for calling schema registry and clarify if the plugin need to test against Kafka v3 and v4

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+     ensure_readable_and_non_writable! "ssl_key", @ssl_key if @ssl_key
+
+    raise_config_error! "`ssl_keystore_path` requires `ssl_keystore_password`" if @ssl_keystore_path && !@ssl_keystore_password
+    raise_config_error! "`ssl_keystore_password` is not allowed unless `ssl_keystore_path` is specified" if @ssl_keystore_password && !@ssl_keystore_path
```
The password for keystore and truststore should be optional. 
In [es-output](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/main/lib/logstash/outputs/elasticsearch/http_client_builder.rb#L173) and [kafka-integration](https://github.com/logstash-plugins/logstash-integration-kafka/blob/main/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb#L98), they are optional

mashhurs replied:
From security point of view, it is a good practise to protect the key/trust stores. In thew `elastic_integration`, when we designed SSL with Rye, we agreed on the point where path should read-only (protect from file replacements) and protect to steal content (with password). I. would actually to support keeping high standard as security is one of the biggest concern area. And we can anytime open if this is something bothers users alot.
WDYT?

kaisecheng replied:
IMO, having password is best practise but users should still be allowed to use keystore without password. Especially, the vast majority of plugins, as well as Logstash and the kafka-input itself, allow no password. I would keep avro-codec aligned with kafka-input

mashhurs replied:
Sounds good. I have opted out the password requirements for both keystore and trustore paths.

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -112,6 +128,23 @@ Use `base64` (default) to indicate that this codec sends or expects to receive b
+* There is no default value for this setting.
+
+Password for HTTP basic authentication when fetching the schema from a remote server.
+Must be used in combination with the `username` setting.
```
```suggestion
Password for HTTP basic authentication when fetching remote schemas.
Used together with `username`.
```

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -112,6 +128,23 @@ Use `base64` (default) to indicate that this codec sends or expects to receive b
+* Value type is <<uri,uri>>
+* There is no default value for this setting.
+
+Proxy server URL for schema registry connections.
```
```suggestion
The address of a forward HTTP proxy to use when contacting a remote schema registry.
```

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,177 @@ example:
+
+Path to PEM encoded certificate file for client authentication (mutual TLS).
+This is an alternative to using <<plugins-{type}s-{plugin}-ssl_keystore_path>>.
+You cannot use this setting and <<plugins-{type}s-{plugin}-ssl_keystore_path>> at the same time.
```
How about this?

```suggestion
You may use this setting or <<plugins-{type}s-{plugin}-ssl_keystore_path>>, but not both simultaneously.
```

kaisecheng replied:
can you also update ssl_certificate_authorities and ssl_key?

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,177 @@ example:
+The path to the JKS or PKCS12 keystore file for client certificate authentication.
+Use this when the schema registry requires mutual TLS (mTLS) authentication.
+
+This setting is only used when `schema_uri` uses the `https://` scheme.
```
```suggestion
```

All ssl_* settings are only relevant when the URI uses "https://". I think it is too much verbose to mention that here.

mashhurs replied:
Agree!

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,177 @@ example:
+The path to the JKS or PKCS12 truststore file containing certificates to verify
+the schema registry server's certificate.
+
+This setting is only used when `schema_uri` uses the `https://` scheme.
```
```suggestion
```

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,177 @@ example:
+* Value type is <<string,string>>
+* There is no default value for this setting.
+
+The type of the truststore file. Can be either `JKS` or `PKCS12`.
```
```suggestion
The format of the truststore file. It must be either `JKS` or `PKCS12`.
```

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -156,3 +360,26 @@ input {
+* There is no default value for this setting.
+
+Username for HTTP basic authentication when fetching the schema from a remote server.
+Must be used in combination with the `password` setting.
```
```suggestion
Username for HTTP basic authentication when fetching remote schemas.
Used together with `password`.
```

kaisecheng reviewed (commented):
Added some suggestion for docs

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+  end
+
+  def validate_ssl_settings!
+    @ssl_enabled = true if @ssl_enabled.nil?
```
I think we need to raise error when ssl_enabled is set to false

mashhurs replied:
Good catch! This was an intention and I missed.

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+      body
+    end
+  ensure
+    client.close if client
```
suggest to add `client = nil` at the beginning of the function. If there is exception before `client` is declared,  `ensure` block will fail.

mashhurs replied:
Yup, agree!

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +183,165 @@ def encode(event)
+
+    if @username && @password
+      raise LogStash::ConfigurationError, "Empty `username` or `password` is not allowed" if @username.empty? || @password.value.empty?
+    end
```
```suggestion
    raise LogStash::ConfigurationError, "Empty `username` or `password` is not allowed" if @username.empty? || @password.value.empty?
```

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+    Base64.strict_encode64(buffer.string)
+  end
+
+  def create_test_schema_file(filename = "test_schema.avsc")
```
no usage

mashhurs replied:
Yup, removed.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+      let(:full_schema_url) do
+        url = "#{schema_registry_url}/subjects/#{schema_subject}/versions/latest"
+        puts "Constructed schema URL: #{url}"
+        raise "Schema URL is empty!" if url.nil? || url.empty? || !url.start_with?('http')
```
url must be starting with "http", right? We can delete it

mashhurs replied:
Removed.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+        expect {
+          invalid_config = { 'schema_uri' => full_schema_url, 'username' => 'invalid', 'password' => 'wrong' }
+          LogStash::Codecs::Avro.new(invalid_config).tap { |c| c.register }
+        }.to raise_error(/401|403|HTTP request failed/)
```
I would expect 401 Unauthorized. Any reason to expect 403 or "HTTP request failed" ?

mashhurs replied:
Yeah I think being specific seems a better approach.
Kept only 401.

kaisecheng commented on spec/integration/avro_integration_spec.rb:
```diff
@@ -0,0 +1,460 @@
+        end
+
+        expect(events.size).to eq(1)
+        expect(events.first.get("message")).to eq(test_event_data["message"])
```
This part has repeated a couple of times. Can we refactor it to a method?

mashhurs replied:
Yup, simplified even `codec.decode(encoded_data) do |event|...` parts.

kaisecheng reviewed (commented):
Regarding the test, can we add some tests for ssl_keystore_*?

mashhurs committed:
Apply suggestions from code review

Apply code review comments which I agree with.

Co-authored-by: kaisecheng <69120390+kaisecheng@users.noreply.github.com>

mashhurs committed:
Simplifications.

mashhurs committed:
Implement a retry mechanism when hitting the schema registry.

mashhurs committed:
Put back accidental removed piece while testing.

mashhurs committed:
Setup schema registry for mutual TLS and add integration test cases for it.

mashhurs commented:
> Regarding the test, can we add some tests for ssl_keystore_*?

Added! This required to setup a mutual config in schema registry as well. Introduced new schema-registry-mutual.properties config with `confluent.http.server.ssl.client.authentication=REQUIRED` option.

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,172 @@ example:
+
+Path to PEM encoded CA certificate file(s) for server verification.
+This is an alternative to using <<plugins-{type}s-{plugin}-ssl_truststore_path>>.
+You cannot use this setting and <<plugins-{type}s-{plugin}-ssl_truststore_path>> at the same time.
```
You may use this setting or <<plugins-{type}s-{plugin}-ssl_truststore_path>>, but not both simultaneously.

mashhurs replied:
```suggestion
You may use this setting or <<plugins-{type}s-{plugin}-ssl_truststore_path>>, but not both simultaneously.
```

kaisecheng commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,172 @@ example:
+* Value type is <<string,string>>
+* There is no default value for this setting.
+
+The type of the keystore file. Can be either `JKS` or `PKCS12`.
```
```suggestion
The format of the keystore file. It must be either `jks` or `pkcs12`.
```

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -84,9 +87,59 @@ class LogStash::Codecs::Avro < LogStash::Codecs::Base
+  config :ssl_keystore_password, :validate => :password
+
+  # Keystore type (JKS or PKCS12)
+  config :ssl_keystore_type, :validate => %w[JKS PKCS12]
```
it seems es-output uses [lowercase](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/1b84db7d2b41d17ef0ef3d156700e345235a2ebd/lib/logstash/plugin_mixins/elasticsearch/api_configs.rb#L78) jks, pkcs12. Can we align with other plugins? Also, update the docs to lowercase.

mashhurs replied:
Yup, done.

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -84,9 +87,59 @@ class LogStash::Codecs::Avro < LogStash::Codecs::Base
+  config :ssl_truststore_password, :validate => :password
+
+  # Truststore type (JKS or PKCS12)
+  config :ssl_truststore_type, :validate => %w[JKS PKCS12]
```
suggest to use lowercase options

mashhurs replied:
```suggestion
  # Truststore type (jks or pkcs12)
  config :ssl_truststore_type, :validate => %w[jks pkcs12]
```

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +184,170 @@ def encode(event)
+    client = Manticore::Client.new(client_options)
+
+    body = Stud.try(3.times, [Manticore::SocketException, Manticore::Timeout, StandardError]) do
+      @logger.debug("Fetching schema from #{uri_string}") if @logger
```
Does the `if @logger` intend to be here? Or were you doing 

`@logger.debug && @logger.debug("Fetching schema from #{uri_string}")` ?

mashhurs replied:
```suggestion
      @logger.debug("Fetching schema from #{uri_string}")
```

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +184,170 @@ def encode(event)
+
+      unless response.code == 200
+        error_msg = "HTTP request failed: #{response.code} #{response.message}"
+        @logger.warn(error_msg) if @logger
```
logger always exists

mashhurs replied:
```suggestion
        @logger.warn(error_msg)
```

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +184,170 @@ def encode(event)
+      response = client.get(uri_string).call
+
+      unless response.code == 200
+        error_msg = "HTTP request failed: #{response.code} #{response.message}"
```
suggestion: "Failed to fetch schema from #{uri_string}: #{response.code} - #{response.message}"

mashhurs replied:
```suggestion
        error_msg = "Failed to fetch schema from #{uri_string}: #{response.code} - #{response.message}"
```

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -131,4 +184,170 @@ def encode(event)
+
+    client = Manticore::Client.new(client_options)
+
+    body = Stud.try(3.times, [Manticore::SocketException, Manticore::Timeout, StandardError]) do
```
how about catching Manticore::ManticoreException for broader network related exceptions and being more specific for BadResponseCodeError? It requires to create BadResponseCodeError

mashhurs replied:
I couldn't quite catch why we need to introduce `BadResponseCodeError` but I have reshaped the logic similar to es-output that has:
- better visibility in logs: retry, what happened, etc..;
- retry block
- exponential sleep
Let me know if I still can improve.

kaisecheng replied:
The version we were discussing 

```ruby
body = Stud.try(3.times, [Manticore::SocketException, Manticore::Timeout, StandardError]) do
    response = client.get(uri_string).call
    unless response.code == 200
      error_msg = ...
      ...
      raise error_msg
    end
end
```

Catching StandardError could be too broad. The purpose of catching StandardError is to retry non 200 only excluding NoMethodError, TypeError, ArgumentError etc
manticore may throw StandardError. I am not sure. That's why I suggested to narrow down the error type
The suggestion is not a blocker to merge.

mashhurs replied:
Same discussion with - https://github.com/logstash-plugins/logstash-codec-avro/pull/47#discussion_r2564696623

kaisecheng reviewed (commented):
The second look mainly about the schema request handling and keeping the configs align with other plugins. Once they are addressed, I can approve it. 

We need Karen to review the docs.  

Also, if you decide to test against Kafka v4 only, please remove the [pairs](https://github.com/logstash-plugins/logstash-codec-avro/pull/47/files#diff-7b436cb5923ba7f66ccf2e140b342330400dcb1d6776d03444564dc34c5379f1R13) and add a comment "v3 is not supported in test"

kaisecheng commented on spec/unit/resources/do_not_remove_path/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Empty dir for test cases to run, imitates readable/non-readable/writable folder
```
typo: initiates

mashhurs replied:
ah by imitates I mean simulates or emulates...

mashhurs commented on lib/logstash/codecs/avro.rb:
```diff
@@ -84,9 +87,59 @@ class LogStash::Codecs::Avro < LogStash::Codecs::Base
+  config :ssl_keystore_password, :validate => :password
+
+  # Keystore type (JKS or PKCS12)
+  config :ssl_keystore_type, :validate => %w[JKS PKCS12]
```
```suggestion
  # Keystore type (jks or pkcs12)
  config :ssl_keystore_type, :validate => %w[pkcs12 jks]
```

mashhurs commented on docs/index.asciidoc:
```diff
@@ -134,6 +167,172 @@ example:
+* Value type is <<string,string>>
+* There is no default value for this setting.
+
+The format of the truststore file. It must be either `JKS` or `PKCS12`.
```
```suggestion
The format of the truststore file. It must be either `jks` or `pkcs12`.
```

mashhurs commented on .ci/run.sh:
```diff
@@ -0,0 +1,34 @@
+else
+  # Define the Kafka:Confluent version pairs
+  VERSIONS=(
+  # "3.9.1:7.4.0"
```
```suggestion
```

mashhurs committed:
Apply suggestions from code review

Truststore and keystore formats lowercased. Logging improvements.

Co-authored-by: kaisecheng <69120390+kaisecheng@users.noreply.github.com>

mashhurs committed:
Update specs after making keystore and trustore type lowecase.

mashhurs committed:
Allow truststore and keystore without using their password.

mashhurs committed:
Retry mechanism test.

kaisecheng commented on lib/logstash/codecs/avro.rb:
```diff
@@ -226,17 +225,36 @@ def fetch_remote_schema(uri_string)
+    rescue StandardError => e
+      # Don't retry HTTP errors (401, 404, etc.) or other non-transient errors
+      @logger.error("Failed to fetch schema from #{uri_string}: #{e.class} - #{e.message}")
+      raise
```
I am a bit confused now why non 200 need to raise StandardError. We want to retry 5xx and no retry for 4xx.
Here, all 4xx and 5xx are caught and no retry, right?

Also, I would like to reorganize the begin rescue block to improve readability. 

```ruby
def fetch_remote_schema(uri_string)
  client = nil
  client_options = {}
  ...
  max_retries = 3
  retry_count = 0
  
  begin
    response = client.get(uri_string).call
    
    unless response.code == 200
      @logger.error("Failed to fetch schema from #{uri_string}: #{response.code} - #{response.message}")
      raise BadResponseCodeError.new(response.code, response.message, uri_string)
    end
    
    body = response.body
    
    # Parse and extract schema
    parsed = JSON.parse(body)
    parsed.is_a?(Hash) && parsed.has_key?('schema') ? parsed['schema'] : body
  rescue JSON::ParserError
    return body
  rescue Manticore::ManticoreException, BadResponseCodeError => e
    # 4xx don't retry
    if e.is_a?(BadResponseCodeError) && e.code >= 400 && e.code < 500
      @logger.error()
      raise
    end

    # retry block
    retry_count += 1
    if retry_count <= max_retries
      ...
    end
  end
ensure
  client.close if client
end
```

mashhurs replied:
Okay, so my confusion was if we introduce `BadResponseCodeError` this will anyway inherit `StandardError` and es-output has same class (similar approach we are following here) which inherits from `LogStash::Error`. In couple of places where LS interfaces with remote hosts like ES, we use `LogStash::Error`. From this point of view, it makes sense to introduce `BadResponseCodeError` and apply your point.

I have added with this commit - https://github.com/logstash-plugins/logstash-codec-avro/pull/47/commits/86c5421e49f037c16ffd3efe495da0a807ef6c4a

kaisecheng commented:
@mashhurs The only unresolved is the retry logic. Others are good

mashhurs committed:
Introduce BadResponseCodeError class to handle non-retriable cases.

mashhurs committed:
Update retry error message.

kaisecheng reviewed (approved):
LGTM 🚀
